compiler/record-hessian/src/HessianComputer.cpp

```diff
@@ -0,0 +1,196 @@
+  }
+  length = num_elements / size_in_ch;
+
+  std::vector<float> hessian(size_in_ch * size_in_ch, 0);
+
+  for (int i = 0; i < size_in_ch; ++i)
+  {
+    for (int j = 0; j < size_in_ch; ++j)
+    {
+      float sum = 0;
+      for (int k = 0; k < length; ++k)
+      {
+        sum += buf[i + k * size_in_ch] * buf[j + k * size_in_ch];
+      }
+      hessian[i * size_in_ch + j] = 2 * sum;
+    }
+  }
+
+  HessianVector &vector = _hessian_map[node];
+  vector.update(hessian);
+}
+
+void HessianComputer::recordHessianForConv2D(const luci::CircleNode *node,
+                                             const luci_interpreter::Tensor *input_tensor)
+{
+  const auto node_filter = loco::must_cast<luci::CircleConst *>(
+    loco::must_cast<const luci::CircleConv2D *>(node)->filter());
+  const auto node_bias =
+    loco::must_cast<luci::CircleConst *>(loco::must_cast<const luci::CircleConv2D *>(node)->bias());
+  uint32_t size_in_ch =
+    node_filter->size<loco::DataType::FLOAT32>() / node_bias->size<loco::DataType::FLOAT32>();
+
+  uint32_t input_n = input_tensor->shape().dim(0);
+  uint32_t input_h = input_tensor->shape().dim(1);
+  uint32_t input_w = input_tensor->shape().dim(2);
+  uint32_t input_c = input_tensor->shape().dim(3);
+
+  uint32_t stride_h = ((luci::CircleConv2D *)node)->stride()->h();
+  uint32_t stride_w = ((luci::CircleConv2D *)node)->stride()->w();
+  uint32_t dilation_h = ((luci::CircleConv2D *)node)->dilation()->h();
+  uint32_t dilation_w = ((luci::CircleConv2D *)node)->dilation()->w();
+
+  uint32_t kernel_oc = node_filter->dim(0).value();
+  uint32_t kernel_h = node_filter->dim(1).value();
+  uint32_t kernel_w = node_filter->dim(2).value();
+  uint32_t kernel_ic = node_filter->dim(3).value();
+
+  const auto data = input_tensor->data<float>();
+  const auto num_elements = input_tensor->shape().num_elements();
+  std::vector<float> buf(data, data + num_elements);
+
+  unfold(buf, input_n, input_h, input_w, input_c, stride_h, stride_w, dilation_h, dilation_w,
+         kernel_oc, kernel_h, kernel_w, kernel_ic);
+  uint32_t length = buf.size() / size_in_ch;
+
+  std::vector<float> hessian(size_in_ch * size_in_ch, 0);
+  for (int i = 0; i < size_in_ch; ++i)
+  {
+    for (int j = 0; j < size_in_ch; ++j)
+    {
+      float sum = 0;
+      for (int k = 0; k < length; ++k)
+      {
+        sum += buf[i + k * size_in_ch] * buf[j + k * size_in_ch];
+      }
+      hessian[i * size_in_ch + j] = 2 * sum;
+    }
+  }
+
+  HessianVector &vector = _hessian_map[node];
+  vector.update(hessian);
+}
+
+void HessianComputer::recordHessian(const luci::CircleNode *node,
+                                    const luci_interpreter::Tensor *input_tensor)
+{
+  if (node == nullptr || input_tensor == nullptr)
+  {
+    throw std::invalid_argument("node or input_tensor is null.");
+  }
+
+  if (input_tensor->element_type() != loco::DataType::FLOAT32)
+  {
+    throw std::runtime_error("Unsupported dtype: only FLOAT32 is supported.");
+  }
```

Comment:
```suggestion
  if (input_tensor->element_type() != loco::DataType::FLOAT32)
    throw std::runtime_error("Unsupported dtype: only FLOAT32 is supported.");
```

Reply:
it would be better to add this tool name in the message so that users can identify which program has thrown the mssage.